sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PreparePubsubWriteDoFn.java

```diff
@@ -165,6 +182,14 @@ public void process(
         return;
       }
     }
+    if (!allowOrderingKey && message.getOrderingKey() != null) {
+      badRecordRouter.route(
+          o,
+          element,
+          inputCoder,
+          null,
+          "The transform was not configured to publish messages with ordering keys");
+    }
```

Comment:
The legacy coders (which are needed for compatibility) silently drop the ordering key during GBK, so letting them through would result in unexpected behavior and potential data loss.
Same for Windmill's current implementation, which is why the transform replacement was modified to fail explicitly.

Reply:
Relaxed this clause a bit while cleaning mixed use of `null` and `""` ordering keys (equivalent to Pub/Sub), but this is still needed to allow recovery via DLQ in case a sink (only Dataflow iirc) omits the ordering key on publish.
Added a TODO above this to follow up once every sink supports ordering key writes.

Reply:
I see. Will the Dataflow sink send these messages be sent to the DLQ too?

Can we add a clear message in CHANGES.md to give users a heads up about this new behavior?

Reply:
The Dataflow sink override translator will throw an exception if `allowOrderingKey` is set, see the change in DataflowRunner.java.

I'll add a line in CHANGES.md, if users were implicitly or explicitly depending on the current behavior which silently drops the ordering key on every message then this change in behavior should be clearly communicated. 👍

Reply:
Done.